crates/nargo_toml/src/lib.rs

```diff
@@ -18,48 +17,84 @@ mod git;
 pub use errors::ManifestError;
 use git::clone_git_repo;
 
-/// Returns the path of the root directory of the package containing `current_path`.
+/// Returns the [Path] of the directory containing the `Nargo.toml` by searching from `current_path` to the root of its [Path].
 ///
-/// Returns a `CliError` if no parent directories of `current_path` contain a manifest file.
+/// Returns a [ManifestError] if no parent directories of `current_path` contain a manifest file.
 pub fn find_package_root(current_path: &Path) -> Result<PathBuf, ManifestError> {
-    let manifest_path = find_package_manifest(current_path)?;
+    let root = path_root(current_path);
+    let manifest_path = find_package_manifest(&root, current_path)?;
 
     let package_root =
         manifest_path.parent().expect("infallible: manifest file path can't be root directory");
 
     Ok(package_root.to_path_buf())
 }
 
-/// Returns the path of the manifest file (`Nargo.toml`) of the package containing `current_path`.
-///
-/// Returns a `CliError` if no parent directories of `current_path` contain a manifest file.
-pub fn find_package_manifest(current_path: &Path) -> Result<PathBuf, ManifestError> {
-    current_path
-        .ancestors()
-        .find_map(|dir| find_file(dir, "Nargo", "toml"))
-        .ok_or_else(|| ManifestError::MissingFile(current_path.to_path_buf()))
-}
+// TODO: We are probably going to need a "filepath utils" crate soon
+fn path_root(path: &Path) -> PathBuf {
```

Comment:
```rust
fn path_root(path: &Path) -> PathBuf {
    let mut components = path.components();

    match (components.next(), components.next()) {
        // Preserve prefix if one exists
        (Some(prefix @ Component::Prefix(_)), Some(root_component @ Component::RootDir)) => {
            PathBuf::from(prefix.as_os_str()).join(root_component.as_os_str())
        }
        (Some(root_component @ Component::RootDir), _) => PathBuf::from(root_component.as_os_str()),

        _ => PathBuf::new(),
    }
}
```

This is equivalent, no? I think the current implementation is a bit odd with how it iterates over all components when it knows that it's going to take 2 at most.

Reply:
I believe that's wrong because workspace members can be specified at any depth:
```toml
[workspace]
members = ["crates/foo/bar/baz", "some/other/path"]
```
And we want to traverse to the workspace file.